PSReadLine/Render.cs

```diff
@@ -242,12 +242,128 @@ private void Render()
+
+            static int FindCommonPrefixLength(string leftStr, string rightStr)
+            {
+                if (string.IsNullOrEmpty(leftStr) || string.IsNullOrEmpty(rightStr))
+                {
+                    return 0;
+                }
+
+                int i = 0;
+                int minLength = Math.Min(leftStr.Length, rightStr.Length);
+
+                while (i < minLength && leftStr[i] == rightStr[i])
+                {
+                    i++;
+                }
+
+                return i;
+            }
+
+            // For screen readers, we are just comparing the previous and current buffer text
+            // (without colors) and only writing the differences.
+            string currentBuffer = ParseInput();
+            string previousBuffer = _previousRender.lines[0].Line;
+
+            // In case the buffer was resized.
+            RecomputeInitialCoords(isTextBufferUnchanged: false);
+
+            // Make cursor invisible while we're rendering.
+            _console.CursorVisible = false;
+
+            // Calculate what to render and where to start the rendering.
+            // TODO: Short circuit optimization when currentBuffer == previousBuffer.
+            int commonPrefixLength = FindCommonPrefixLength(previousBuffer, currentBuffer);
+
+            if (commonPrefixLength > 0 && commonPrefixLength == previousBuffer.Length)
+            {
+                // Previous buffer is a complete prefix of current buffer.
+                // Just append the new data.
+                var appendedData = currentBuffer.Substring(commonPrefixLength);
+                _console.Write(appendedData);
+            }
+            else if (commonPrefixLength > 0)
+            {
+                // Buffers share a common prefix but previous buffer has additional content.
+                // Move cursor to where the difference starts, clear forward, and write the data.
+                var diffPoint = ConvertOffsetToPoint(commonPrefixLength);
+                _console.SetCursorPosition(diffPoint.X, diffPoint.Y);
+                var changedData = currentBuffer.Substring(commonPrefixLength);
+                _console.Write("\x1b[0J");
+                _console.Write(changedData);
+            }
+            else
+            {
+                // No common prefix, rewrite entire buffer.
+                _console.SetCursorPosition(_initialX, _initialY);
+                _console.Write("\x1b[0J");
+                _console.Write(currentBuffer);
+            }
+
+            // If we had to wrap to render everything, update _initialY
+            var endPoint = ConvertOffsetToPoint(currentBuffer.Length);
+            int physicalLine = endPoint.Y - _initialY;
+            if (_initialY + physicalLine > bufferHeight)
+            {
+                // We had to scroll to render everything, update _initialY.
+                _initialY = bufferHeight - physicalLine;
+            }
```

Comment:
Y is in the range [0, buffer_height), so the check should be `>= bufferHeight`
We should calculate how many physical lines were scrolled up, and move the initial Y up the same number of lines.
```suggestion
            if (endPoint.Y >= bufferHeight)
            {
                // We had to scroll to render everything, update _initialY.
                int scolledLines = endPoint.Y - bufferHeight + 1;
                _initialY -= scolledLines;
            }
```

Reply:
There is one edge case that I can think of: `endPoint.X == 0 && endPoint.Y == bufferHeight`. In such a case, there is no scrolling yet -- the written-out text exactly fits to the last physical line, and thus when calculating the cursor position, we get (0, bufferHeight). In this case, we shouldn't update `_initialY`.

Reply:
Hm, I'd want to test this change a bunch. As is, I'm relying on the existing logic being proven. If we change it here, do we want to change it in `ReallyRender` too?